src/Language/Futhark/Interpreter/Values.hs

```diff
@@ -76,32 +76,21 @@ emptyShape :: ValueShape -> Bool
 emptyShape (ShapeDim d s) = d == 0 || emptyShape s
 emptyShape _ = False
 
-typeShape :: M.Map VName (Shape d) -> TypeBase d () -> Shape d
-typeShape shapes = go
-  where
-    go (Array _ _ shape et) =
-      foldr ShapeDim (go (Scalar et)) $ shapeDims shape
-    go (Scalar (Record fs)) =
-      ShapeRecord $ M.map go fs
-    go (Scalar (Sum cs)) =
-      ShapeSum $ M.map (map go) cs
-    go (Scalar (TypeVar _ _ (QualName [] v) []))
-      | Just shape <- M.lookup v shapes =
-          shape
-    go _ =
-      ShapeLeaf
+typeShape :: TypeBase d () -> Shape d
+typeShape (Array _ _ shape et) =
+  foldr ShapeDim (typeShape (Scalar et)) $ shapeDims shape
+typeShape (Scalar (Record fs)) =
+  ShapeRecord $ M.map typeShape fs
+typeShape (Scalar (Sum cs)) =
+  ShapeSum $ M.map (map typeShape) cs
+typeShape _ =
+  ShapeLeaf
 
-structTypeShape :: M.Map VName ValueShape -> StructType -> Shape (Maybe Int64)
-structTypeShape shapes = fmap dim . typeShape shapes'
+structTypeShape :: StructType -> Shape (Maybe Int64)
+structTypeShape = fmap dim . typeShape
   where
-    dim (SizeExpr (IntLit d _ _)) = Just $ fromIntegral d
+    dim (SizeExpr (Literal (SignedValue (Int64Value d)) _)) = Just $ fromIntegral d
```

Comment:
Current implementation of sizes tends to prefer `IntLit` to `Literal` for pattern matching, notably `sizeFromInteger` which replace the old `ConstSize` will make `SizeExpr (IntLit ...)`, because `1` is parsed as an `IntLit`.
The fact that you have some seems to come from an error I've made in `evalType.matchPtoA` at [line 576-577](https://github.com/diku-dk/futhark/blob/arbitrary-size-expressions/src/Language/Futhark/Interpreter.hs#L576-577), at first I used `Literal` not knowing where which one to choose and I forgot this occurence

Reply:
I will change this to using `IntLit` consistently.